File: 0046-permutations/0046-permutations.py
class Solution(object):
    def permute(self, nums):
        def backtrack(start,end):
            if start==end:
                res.append(nums[:])
                return
            for i in range(start,end):
                nums[start],nums[i]=nums[i],nums[start]
                backtrack(start+1,end)
                nums[start],nums[i]=nums[i],nums[start]
        res=[]
        backtrack(0,len(nums))
        return res

File: 0046-permutations/0046-permutations.java
import java.util.*;

class Solution {
    public List<List<Integer>> permute(int[] nums) {
        List<List<Integer>> res=new ArrayList<>();
        backtrack(0,nums.length,res,nums);
        return res;
    }

    private static void backtrack(int start,int end,List<List<Integer>> res,int[] nums){
        if(start == end){
            List<Integer> list = new ArrayList<>();
            for (int num : nums) {
                list.add(num);
            }
            res.add(list);
            return;
        }
        for(int i=start; i<end; i++){
            rev(start,i,nums);
            backtrack(start+1,end,res,nums);    
            rev(start,i,nums);    
        }
    }

    private static void rev(int i,int start,int[] nums){
        int temp=nums[start];
        nums[start]=nums[i];
        nums[i]=temp;
    }
}